### ISSUE TITLE
[v0.10] When two characters are ligatured, the first one becomes invisible

### ISSUE THREAD

**Issue by xa0082249956** [microsoft/terminal#4974]
![fig2](https://user-images.githubusercontent.com/9026813/76924254-bc6ee980-6910-11ea-8391-fc5ab1d9fe02.gif)

## Environment

```none
Platform ServicePack Version      VersionString
-------- ----------- -------      -------------
 Win32NT             10.0.19582.0 Microsoft Windows NT 10.0.19582.0
```

## Steps to reproduce

1. Change font to the one with ligatures, for example *Fira Code*.
2. Type `--`.

## Expected behavior

The first `-` shouldn't disappear.

## Actual behavior

The first `-` is disappeared.

**Comment by DHowett-MSFT** [comment]
What font size are you using? What version of Terminal?

**Comment by xa0082249956** [comment]
> What font size are you using? 

Default size i.e. 12.

> What version of Terminal?

Windows Terminal (Preview)
Version: 0.10.761.0
Downloaded from Microsoft Store.

**Comment by onekopaka** [comment]
I am also having this issue, using 10pt Fira Code, no matter the zoom level, I just built from commit c0d704e734c54381af59dc0413b7aa19326a84ed

Here you can see the //, :: and www ligatures drawing incorrectly:
![Screenshot of my terminal, showing the www, :: and // ligatures drawing incorrectly](https://ss.theoks.net/WindowsTerminal_2020-03-18_02-47-59.png)

**Comment by DHowett-MSFT** [comment]
We thought we'd fixed this with #4806, but perhaps we didn't.

**Comment by allenatdecisiv** [comment]
Same here:
![terminal-preview](https://user-images.githubusercontent.com/1952547/77141979-15479a80-6a55-11ea-9c92-fe8df08ac9bf.gif)

`            "fontFace" : "Fira Code Retina",`
`            "fontSize" : 12,`
`Windows Terminal (Preview)`
`Version: 0.10.761.0`
`Microsoft Windows [Version 10.0.18363.720]`

**Comment by daviox** [comment]
Same here:
![Ligatures issue](https://user-images.githubusercontent.com/25099832/77161606-646bea80-6aaa-11ea-9856-5bd6999d14a6.png)

### Versions
> Windows Terminal (Preview)
> Version: 0.10.761.0

> Microsoft Windows [Version 10.0.18363.719]

### Config
```json
      "fontFace": "Fira Code",
      "fontSize": 9,
      "scrollbarState": "hidden",
      "padding": "5",
```

**Comment by fabienheureux** [comment]
do you have an ETA for this issue @DHowett-MSFT ? 
Thanks 👍

**Comment by DHowett-MSFT** [comment]
Unfortunately, I do not. Sorry.

**Comment by fabienheureux** [comment]
I can add to this issue that switching to Cascadia Code fixes this issue (I was using Fira Code).

**Comment by skyline75489** [comment]
I'm using WT Version 0.10.781.0 and I can not seem to reproduce it. I'm also a Fira Code lover myself.

**Comment by FallenGameR** [comment]
I'm on WT 0.10.781.0 and do reproduce this bug with Fira Code.

**Comment by zadjii-msft** [comment]
I'd bet this might have to do with not only the font face, but also the font size and display resolution. @FallenGameR (or anyone else reproing this), could you also make sure to share your `fontSize` and display resolution? Thanks!

**Comment by fabienheureux** [comment]
@zadjii-msft  here are the settings that causes this issue: 
```
"fontFace" : "Fira Code",
"fontSize" : 10,
```

Resolution: 1920 x 1080

**Comment by anttirt** [comment]
Reproduced with Microsoft Terminal 0.10.781.0, Fira Code 1.206, display resolution 1920x1200, every font size from 6 to 14.

Updating to Fira Code 2.000 fixed the issue and all ligatures show up correctly at all settings I'm able to test.

**Comment by skyline75489** [comment]
Ah. That's probably it. I'm at 2560 x 1600. I can't find a way to reproduce it.

**Comment by fabienheureux** [comment]
> 
> 
> Reproduced with Windows Terminal 0.10.781.0, Fira Code 1.206, display resolution 1920x1200, every font size from 6 to 14.
> 
> Updating to Fira Code 2.000 fixed the issue and all ligatures show up correctly at all settings I'm able to test.

The update to Fira Code 2.000 fixed the issue as well on my side

**Comment by griimick** [comment]
Same. Updating to Fira Code v.2 fixed the issues on my side.

**Comment by neillrobson** [comment]
Updating to v2 fixed the issue for me as well. Just FYI: make sure that when you are updating (or checking your font version numbers), you are looking at _all_ of the font faces in the settings page. I am embarrassed to say that, somehow, I had installed v2.0 of Light and Bold, while keeping version 1.2 around for Regular, Medium, and Retina. No idea how that happened. Just to be safe, I uninstalled the Chocolatey package that I had been using to automatically update the font face.

**Comment by allenatdecisiv** [comment]
Updating to Fira Code 2 fixed the issue for me as well. I did have to manually remove all Fira Code fonts, then install the new ones. Overwriting didn't fix it.
![fira-code-2-works](https://user-images.githubusercontent.com/1952547/77785794-0e132480-7033-11ea-88e8-2934a423e9be.gif)
```
        {
            "acrylicOpacity" : 0.85,
            "closeOnExit" : true,
            "colorScheme" : "Campbell",
            "commandline" : "cmd.exe",
            "cursorColor" : "#FFFFFF",
            "cursorShape" : "filledBox",
            "fontFace" : "Fira Code Retina",
            "fontSize" : 12,
            "guid" : "{0caa0dad-35be-5f56-a8ff-afceeeaa6101}",
            "historySize" : 9001,
            "icon" : "ms-appx:///ProfileIcons/{0caa0dad-35be-5f56-a8ff-afceeeaa6101}.png",
            "name" : "Command Prompt",
            "padding" : "0, 0, 0, 0",
            "snapOnInput" : true,
            "startingDirectory" : "%USERPROFILE%",
            "useAcrylic" : true
        },
```

**Comment by FallenGameR** [comment]
Yep. Updating Fira Code to v2 fixed it. On one machine fonts were able to update the old version of themselves. On another machine v2 said that it was successfully installed, but v1.2 still remained and was picked up by default. Had to reboot and manually delete v1.2.

> I'd bet this might have to do with not only the font face, but also the font size and display resolution. @FallenGameR (or anyone else reproing this), could you also make sure to share your fontSize and display resolution? Thanks!

2560x1600 - machine A
3240x2160 - machine B, 200% font scale
"fontSize": 14 - on both machines

**Comment by daviox** [comment]
Updating FiraCode to v2 solved the issue. Thanks!
![No issues here](https://user-images.githubusercontent.com/25099832/77826943-1478cd80-7113-11ea-8311-3a95a8ca765f.gif)

**Comment by privacyguy123** [comment]
Updating the font fixed this yeah - doesn't seem to be a Terminal bug then?

**Comment by DHowett-MSFT** [comment]
It's still a Terminal bug, unfortunately. Thanks :smile:

**Comment by fabienheureux** [comment]
> It's still a Terminal bug, unfortunately. Thanks 😄

If It can be of any help...this worked before #4806

**Comment by DHowett-MSFT** [comment]
Thanks! 4806 describes the same bug as this, so it definitely worked before 4806 ;)

**Comment by vincentkam** [comment]
I thought I'd share my experience with this bug: I had an older version of the Firacode fonts hanging out in `%APPDATA%\Local\Microsoft\Windows\Fonts` that was being preferentially loaded instead of the 2.0 fonts I had installed into `C:\Windows\Fonts`. I was unable to delete the fonts in `%APPDATA%\Local\Microsoft\Windows\Fonts` despite using elevated command prompts, trying to close all handles to the fonts using Process Explorer (trying to do so resulted in invalid handle messages), and many other methods. The only thing that worked for me was using IObit Unlocker. After removing those font files, Terminal correctly picked up the newer FiraCode font and seems to be working normally now.

Windows 10 Version 1909 (OS Build 1836.752)
Terminal version:  0.10.781.0

**Comment by privacyguy123** [comment]
Any ETA on this?

**Comment by zadjii-msft** [comment]
@privacyguy123 Unfortunately not at this time. I tried reproing this Friday, but I couldn't get this to happen on my machine at all with Fira Code 1.2. Hopefully someone else on the team will have a better chance at reproing this. 

If we can't get a repro locally, we could try psychic debugging by looking at the changelog of what Fira Code changed on their side between 1.2 and 2.0. Maybe that would point us in the right direction of something we've done wrong. 

It's also just possible that some of the rendering changes that are baking for 0.11 (#778/#5185) might just fix this, which might explain why I couldn't get a repro.

**Comment by jsoref** [comment]
@zadjii-msft / @DHowett-MSFT : the bot can't complain about issue titles, but can you please fix `invisible`? 😉

**Comment by DHowett-MSFT** [comment]
@jsoref done

**Comment by DHowett-MSFT** [comment]
:trollface:

**Comment by jsoref** [comment]
Technically, I suppose, I should say "it doesn't". It probably _could_.

**Comment by privacyguy123** [comment]
Come on boys - new release and this was overlooked? It worked fine in earlier versions ...

**Comment by zadjii-msft** [comment]
* This is a purely graphical bug (the characters are still in the buffer, just not rendered for some reason)
* With a trivial workaround (Upgrade the font, or pick _any other font_)
* And the team doesn't have a local repro for this with the FiraCode v1.2 that other users are reporting an error with

So no, this hasn't really bubbled up in our own priority list. Also, at least I was hoping that #5345 would just magically fix this (as there was quite a bit of renderer changes in that PR).

**Comment by privacyguy123** [comment]
I wish I understood how to fix it. 😛

**Comment by illustrious-you** [comment]
This bug occurred on my machine with FiraCode 4.0 at all font sizes. The problem was resolved when I **unblocked the FiraCode ttf files** before installing them.

**Comment by zadjii-msft** [comment]
I'm curious if this just went away in 1.16. The Atlas engine might have just fixed this.

**Comment by DHowett** [comment]
Alright, the last update here was in 2023. I'm calling it good - folks who are seeing this in 1.23 can file a new bug. :)

**Comment by jsoref** [comment]
@DHowett fwiw, the next version of the bot could complain about titles for PRs (it won't do issues anytime soon as I'll be collecting feedback on the general feature and don't have any imminent plans you extend it to issues although I certainly could).

**Comment by DHowett** [comment]
oh, very cool. I do like having competent PR titles. :)
